Operator name separate from class

norm opened the pull request:
It should be more helpful to display '@task.virtualenv' rather than
'_PythonVirtualenvOperator' when looking at dag run details.

Closes #22762.

uranusjr commented on airflow/decorators/python.py:
```diff
@@ -55,6 +55,10 @@ def __init__(self, *, python_callable, op_args, op_kwargs, **kwargs) -> None:
 
+    @property
+    def get_operator_ui_name(self):
+        return 'PythonOperator'
```
I’m thinking maybe we should let this return `"@task.python"`; does that look better in the UI?

norm replied:
@ashb mentioned metrics to me as a reason not to choose that.

uranusjr replied:
But the metrics are currently using `task_type`. Since we’re introducing a new property (like in this PR), we can do whatever we want.

ashb replied:
Yes, TP is right.

norm committed:
Implement custom operator names

It should be more helpful to display '@task.virtualenv' rather than
'_PythonVirtualenvOperator' when looking at dag run details.

uranusjr commented on airflow/decorators/python_virtualenv.py:
```diff
@@ -44,6 +44,8 @@ class _PythonVirtualenvDecoratedOperator(DecoratedOperator, PythonVirtualenvOper
     # there are some cases we can't deepcopy the objects (e.g protobuf).
     shallow_copy_attrs: Sequence[str] = ('python_callable',)
 
+    operator_name: str = 'PythonVirtualenvOperator'
```
Is this `@task.virtualenv`?

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -700,6 +700,8 @@ class derived from this one results in the creation of a task object,
     # Set to True for an operator instantiated by a mapped operator.
     __from_mapped = False
 
+    operator_name: str = 'BaseOperator'
```
Maybe a better default would be `operator_name = ""`?

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1345,6 +1347,10 @@ def get_direct_relatives(self, upstream: bool = False) -> Iterable["DAGNode"]:
+    @property
+    def get_operator_name(self):
+        return self.operator_name
+
```
... and we could implement this as


```python
def get_operator_name(self) -> str:
    return self.operator_name or type(self).__name__
```

p.s. when I use `property` I’d drop `get_`, and vice versa.

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -1458,6 +1466,7 @@ def get_serialized_fields(cls):
                     'start_date',
                     'end_date',
                     '_task_type',
+                    '_operator_name',
```
I don't think this is needed anymore, as operator_name is now a property, not a plain attribute?

norm replied:
I followed how `_task_type` is handled, which is why it was there. Now, if I take it out, `tests/serialization/test_dag_serialization.py::test_operator_expand_serde` fails. I'll have a further dig into it.

ashb replied:
Oh, `_operator_name` is on MappedOperator, so rather than adding this here, add that in to the same named function in MappedOperator instead.

norm replied:
I'm not sure I understand your suggestion — `_operator_name` is the same trick as `_task_type` to preserve the name of *any* operator during serialisation. It is not specific to one?

norm replied:
(and tests fail if I move it as suggested, although that could be that the test is doing the wrong thing)

ashb replied:
On my (admittedly somewhat cursory) reading of this code, a BaseOperator should never have an attribute called `_operator_name`, that only exists as a field on `MappedOperator` instances. Have I read it wrong?

ashb commented on airflow/operators/empty.py:
```diff
@@ -29,6 +29,7 @@ class EmptyOperator(BaseOperator):
 
     ui_color = '#e8f7e4'
     inherits_from_empty_operator = True
+    custom_operator_name = 'Empty'
```
Only remaining question is if we want this (or should we leave it as `EmptyOperator`)

norm replied:
Ah, yes, I did that when updating the tests to prove I wasn't just accidentally getting `task_type`. I'll revert that and use a specific class that's just in the tests.

ashb reviewed (approved):
One small Q.

And a bonus change/nit/follow-up-pr: Don't include `_operator_name` in the serialized output if it's the same value as `_task_type` (to reduce the size of the JSON payload)

eladkal reviewed (approved):
I would suggest to mention this ability in the docs of creating custom operator:
https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html

norm committed:
fixup! Implement custom operator names

norm committed:
fixup! Implement custom operator names

norm committed:
fixup! Implement custom operator names
